drivers/flash/flash_bflb.c

```diff
@@ -0,0 +1,1564 @@
+
+		/* copy data we need */
+		flash_bflb_xip_memcpy((uint8_t *)(address + BFLB_XIP_BASE),
+				      (uint8_t *)buffer, length);
+
+		sys_cache_data_flush_and_invd_all();
+
+		flash_bflb_set_offset(data->reg_copy, img_offset);
+	} else {
+		/* copy data we need */
+		flash_bflb_xip_memcpy((uint8_t *)(address + BFLB_XIP_BASE - img_offset),
+				      (uint8_t *)buffer, length);
+	}
+
+	/* done with interrupt breaking stuffs */
+	irq_unlock(locker);
+
+	return 0;
+}
+
+#endif
+
+static int flash_bflb_write(const struct device *dev,
+			     off_t address,
+			     const void *buffer,
+			     size_t length)
+{
+	struct flash_bflb_data *data = dev->data;
+	uint32_t	tmp;
+	unsigned int	locker;
+	int		ret;
+	uint32_t	cur_len, i;
+	struct bflb_flash_command write_cmd = {0};
+
+	if (length == 0) {
+		return 0;
+	}
+
+	ret = flash_bflb_is_valid_range(address, length);
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (flash_bflb_is_in_xip(&flash_bflb_write)) {
+		return -ENOTSUP;
+	}
+
+	/* interrupting would break, likely to access XIP*/
+	locker = irq_lock();
+
+	ret = flash_bflb_save_xip_state(dev);
+	if (ret != 0) {
+		irq_unlock(locker);
+		return ret;
+	}
+
+	/* Check if the flash chip is OK to write to
+	 * TODO: Ability to disable this and similar when used with things that arent flashes
+	 */
+	ret = flash_bflb_flash_read_register(data, 0, (uint8_t *)(&tmp), 1);
+	if (ret != 0) {
+		goto exit_here;
+	}
+	if ((tmp & BFLB_FLASH_FLASH_BLOCK_PROTECT_MSK) != 0) {
+		ret = -EINVAL;
+		goto exit_here;
+	}
+
+	if ((data->flash_cfg.io_mode & 0xf) == 0
+		|| (data->flash_cfg.io_mode & 0xf) == 1
+		|| (data->flash_cfg.io_mode & 0xf) == 3) {
+		write_cmd.cmd_buf[0] = data->flash_cfg.page_program_cmd << 24;
+	/* quad mode */
+	} else {
+		write_cmd.cmd_buf[0] = data->flash_cfg.qpage_program_cmd << 24;
+		write_cmd.spi_mode =
+			flash_bflb_admode_to_spimode(data->flash_cfg.qpp_addr_mode, 2);
+	}
+	write_cmd.rw = 1;
+	write_cmd.addr_size = 3;
+
+	i = 0;
+	while (i < length) {
+		/* Write enable is needed for every write */
+		ret = flash_bflb_enable_writable(data);
+		if (ret != 0) {
+			goto exit_here;
+		}
+
+		/* Get current position within page size,
+		 * this assumes page_size <= CTRL_BUF_SIZE (512?)
+		 */
+		cur_len = data->flash_cfg.page_size - ((address + i) % data->flash_cfg.page_size);
+
+		if (cur_len > length - i) {
+			cur_len = length - i;
+		}
+
+		flash_bflb_xip_memcpy((uint8_t *)(buffer) +  i, (uint8_t *)SF_CTRL_BUF_BASE,
+				      cur_len);
+		write_cmd.cmd_buf[0] &= ~0xFFFFFF;
+		write_cmd.cmd_buf[0] |= (address + i);
+		write_cmd.nb_data = cur_len;
+		ret = flash_bflb_send_command(data, &write_cmd);
+		if (ret != 0) {
+			goto exit_here;
+		}
+
+		i += cur_len;
+
+		flash_bflb_busy_wait(data);
+		flash_bflb_flash_busy_wait(data);
+	}
+
+	ret = flash_bflb_restore_xip_state(data);
+	irq_unlock(locker);
+
+	return ret;
+
+exit_here:
+	flash_bflb_restore_xip_state(data);
+	irq_unlock(locker);
+
+	return ret;
```

Comment:
This looks ugly and duplicated.
Can it be something like:
```suggestion
exit_here:
	ret2 = flash_bflb_restore_xip_state(data);
	irq_unlock(locker);

	return (ret != 0 ? ret : ret2;
```